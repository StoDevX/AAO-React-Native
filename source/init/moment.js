// @flow
import moment from 'moment-timezone'

// These next values made possible by
// http://stackoverflow.com/questions/42513276

// Tell Moment to round up, so that Moment will always report the
// time until something else in a way that matches the clock time.
// That is, at 5:22pm, Moment will report "8 minutes until 5:30pm"
// until the clock ticks over to 5:23pm.
// If we do Math.floor, we go the other direction: it will report
// "7 minutes until 5:30pm" as soon as 5:22pm starts.
moment.relativeTimeRounding(Math.ceil)

// Tell Moment that a minute is 60 seconds, for reporting purposes.
// It defaults to 45 seconds.
moment.relativeTimeThreshold('s', 60)
