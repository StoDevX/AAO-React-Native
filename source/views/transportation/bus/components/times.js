// @flow

import * as React from 'react'
import {Text} from 'react-native'
import type {DepartureTimeList} from '../types'

const TIME_FORMAT = 'h:mma'

type Props = {|
  +times: DepartureTimeList,
  +style?: any,
|}

export class ScheduleTimes extends React.PureComponent<Props, void> {
  render() {
    const {times} = this.props

    return (
      <Text>
        {times
          // and format the times
          .map(time => (!time ? 'None' : time.format(TIME_FORMAT)))
          .join(' â€¢ ')}
      </Text>
    )
  }
}
