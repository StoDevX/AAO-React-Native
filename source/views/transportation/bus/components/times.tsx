import * as React from 'react'
import {StyleProp, Text, TextStyle} from 'react-native'

import type {DepartureTimeList} from '../types'

const TIME_FORMAT = 'h:mma'

type Props = {
	times: DepartureTimeList
	style: StyleProp<TextStyle>
}

export function ScheduleTimes({times, style}: Props): JSX.Element {
	return (
		<Text style={style}>
			{times
				// and format the times
				.map((time) => time?.format(TIME_FORMAT) ?? 'None')
				.join(' â€¢ ')}
		</Text>
	)
}
