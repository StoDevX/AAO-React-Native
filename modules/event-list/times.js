import moment from 'moment-timezone'
import type {EventType} from '@frogpond/event-type'

export function times(event: EventType) {
	let eventLength = moment
		.duration(event.endTime.diff(event.startTime))
		.asHours()

	let allDay = eventLength === 24
	let multiDay = event.startTime.dayOfYear() !== event.endTime.dayOfYear()
	let sillyZeroLength = event.startTime.isSame(event.endTime, 'minute')

	let startTimeFormatted = event.startTime.format('h:mm A')
	let endTimeFormatted = event.endTime.format('h:mm A')
	let midnightTime = '12:00 AM'

	let start, end
	if (event.isOngoing) {
		start = event.startTime.format('MMM. D')
		end = event.endTime.format('MMM. D')
	} else if (multiDay) {
		// 12:00 PM to Jun. 25 3:00pm
		// Midnight to Jun. 25 <-- assuming the end time is also midnight
		start = startTimeFormatted
		let endFormat =
			endTimeFormatted === midnightTime ? 'MMM. D' : 'MMM. D h:mm A'
		end = `to ${event.endTime.format(endFormat)}`
	} else if (sillyZeroLength) {
		start = startTimeFormatted
		end = 'until ???'
	} else {
		start = startTimeFormatted
		end = endTimeFormatted
	}

	start = start === midnightTime ? 'Midnight' : start
	end = end === midnightTime ? 'Midnight' : end

	return {start, end, allDay}
}

export function detailTimes(event: EventType) {
	let eventLength = moment
		.duration(event.endTime.diff(event.startTime))
		.asHours()

	let allDay = eventLength === 24
	let multiDay = event.startTime.dayOfYear() !== event.endTime.dayOfYear()
	let sillyZeroLength = event.startTime.isSame(event.endTime, 'minute')
	let endsOnSameDay = event.startTime.isSame(event.endTime, 'day')

	let endFormat = endsOnSameDay ? 'h:mm A' : 'MMM. D h:mm A'
	let startTimeFormatted = event.startTime.format('MMM. D h:mm A')
	let endTimeFormatted = event.endTime.format(endFormat)
	let midnightTime = '12:00 AM'

	let start, end
	if (event.isOngoing) {
		start = event.startTime.format('MMM. D')
		end = event.endTime.format('MMM. D')
	} else if (multiDay) {
		// 12:00 PM to Jun. 25 3:00pm
		// Midnight to Jun. 25 <-- assuming the end time is also midnight
		start = startTimeFormatted
		let endFormat =
			endTimeFormatted === midnightTime ? 'MMM. D' : 'MMM. D h:mm A'
		end = `${event.endTime.format(endFormat)}`
	} else if (sillyZeroLength) {
		start = `Starts on ${startTimeFormatted}`
		end = ''
	} else {
		start = startTimeFormatted
		end = endTimeFormatted
	}

	start = start === midnightTime ? 'Midnight' : start
	end = end === midnightTime ? 'Midnight' : end

	return {start, end, allDay}
}
