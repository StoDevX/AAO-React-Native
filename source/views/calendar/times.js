// @flow

import moment from 'moment-timezone'
import type {EventType} from './types'

export function times(event: EventType) {
  const eventLength = moment
    .duration(event.endTime.diff(event.startTime))
    .asHours()

  const allDay = eventLength === 24
  const multiDay = event.startTime.dayOfYear() !== event.endTime.dayOfYear()
  const sillyZeroLength = event.startTime.isSame(event.endTime, 'minute')

  let startTimeFormatted = event.startTime.format('h:mm A')
  let endTimeFormatted = event.endTime.format('h:mm A')
  let midnightTime = '12:00 AM'

  let start, end
  if (event.isOngoing) {
    start = event.startTime.format('MMM. D')
    end = event.endTime.format('MMM. D')
  } else if (multiDay) {
    // 12:00 PM to Jun. 25 3:00pm
    // Midnight to Jun. 25 <-- assuming the end time is also midnight
    start = startTimeFormatted
    const endFormat =
      endTimeFormatted === midnightTime ? 'MMM. D' : 'MMM. D h:mm A'
    end = `to ${event.endTime.format(endFormat)}`
  } else if (sillyZeroLength) {
    start = startTimeFormatted
    end = 'until ???'
  } else {
    start = startTimeFormatted
    end = endTimeFormatted
  }

  start = start === midnightTime ? 'Midnight' : start
  end = end === midnightTime ? 'Midnight' : end

  return {start, end, allDay}
}

export function detailTimes(event: EventType) {
  const eventLength = moment
    .duration(event.endTime.diff(event.startTime))
    .asHours()

  const allDay = eventLength === 24
  const multiDay = event.startTime.dayOfYear() !== event.endTime.dayOfYear()
  const sillyZeroLength = event.startTime.isSame(event.endTime, 'minute')
  const endsOnSameDay = event.startTime.isSame(event.endTime, 'day')

  const endFormat = endsOnSameDay ? 'h:mm A' : 'MMM. D h:mm A'
  let startTimeFormatted = event.startTime.format('MMM. D h:mm A')
  let endTimeFormatted = event.endTime.format(endFormat)
  let midnightTime = '12:00 AM'

  let start, end
  if (event.isOngoing) {
    start = event.startTime.format('MMM. D')
    end = event.endTime.format('MMM. D')
  } else if (multiDay) {
    // 12:00 PM to Jun. 25 3:00pm
    // Midnight to Jun. 25 <-- assuming the end time is also midnight
    start = startTimeFormatted
    const endFormat =
      endTimeFormatted === midnightTime ? 'MMM. D' : 'MMM. D h:mm A'
    end = `${event.endTime.format(endFormat)}`
  } else if (sillyZeroLength) {
    start = `Starts on ${startTimeFormatted}`
    end = ''
  } else {
    start = startTimeFormatted
    end = endTimeFormatted
  }

  start = start === midnightTime ? 'Midnight' : start
  end = end === midnightTime ? 'Midnight' : end

  return {start, end, allDay}
}
