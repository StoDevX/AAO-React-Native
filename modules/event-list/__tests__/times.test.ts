/* eslint-env jest */

import moment from 'moment'
import {times, detailTimes} from '../times.js'

describe('allDay', () => {
	test('should be true for a midnight-to-midnight event', () => {
		const event = {
			startTime: moment('2018-08-07T00:00:00Z'),
			endTime: moment('2018-08-08T00:00:00Z'),
		}
		expect(times(event).allDay).toBe(true)
		expect(detailTimes(event).allDay).toBe(true)
	})

	test('should be true for a noon-to-noon event', () => {
		const event = {
			startTime: moment('2018-08-07T12:00:00Z'),
			endTime: moment('2018-08-08T12:00:00Z'),
		}
		expect(times(event).allDay).toBe(true)
		expect(detailTimes(event).allDay).toBe(true)
	})

	test('should be false for a non-24 hour event', () => {
		const event = {
			startTime: moment('2018-08-07T12:00:00Z'),
			endTime: moment('2018-08-07T12:30:00Z'),
		}
		expect(times(event).allDay).toBe(false)
		expect(detailTimes(event).allDay).toBe(false)
	})
})

xdescribe('ongoing events', () => {
	test('should be formatted', () => {
		const event = {
			startTime: moment().subtract(1, 'hour'),
			endTime: moment().add(1, 'hour'),
		}
		expect(times(event)).toMatchSnapshot()
		// expect(detailTimes(event)).toMatchSnapshot()
	})
})

// multi-day events

// zero-length events

// midnight and noon
// events starting at midnight
// ... ending
// starting ... noon
// ... ending

// unremarkable events

// and they can all test times and detailTimes right next to each other -
// no need for separate testcases.
