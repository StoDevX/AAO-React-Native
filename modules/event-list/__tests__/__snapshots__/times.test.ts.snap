// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ongoing events should be formatted 1`] = `
{
  "allDay": false,
  "end": "Aug. 7",
  "start": "Aug. 7",
}
`;

exports[`ongoing events should be formatted 2`] = `
{
  "allDay": false,
  "end": "Aug. 7",
  "start": "Aug. 7",
}
`;
